src/sentry_core.c

```diff
@@ -264,57 +278,58 @@ event_is_considered_error(sentry_value_t event)
 sentry_uuid_t
 sentry_capture_event(sentry_value_t event)
 {
-    const sentry_options_t *opts = sentry_get_options();
-    if (!opts) {
-        sentry_value_decref(event);
-        return sentry_uuid_nil();
-    }
-    uint64_t rnd;
-    if (opts->sample_rate < 1.0 && !sentry__getrandom(&rnd, sizeof(rnd))
-        && ((double)rnd / (double)UINT64_MAX) > opts->sample_rate) {
-        SENTRY_DEBUG("throwing away event due to sample rate");
-        sentry_value_decref(event);
-        return sentry_uuid_nil();
-    }
+    sentry_envelope_t *envelope = NULL;
+    // we need the options to sample, before_send and attachments
+    SENTRY_WITH_OPTIONS (options) {
+        uint64_t rnd;
+        if (options->sample_rate < 1.0 && !sentry__getrandom(&rnd, sizeof(rnd))
+            && ((double)rnd / (double)UINT64_MAX) > options->sample_rate) {
+            SENTRY_DEBUG("throwing away event due to sample rate");
+            break; // SENTRY_WITH_OPTIONS
+        }
 
-    SENTRY_DEBUG("capturing event");
-    sentry_uuid_t event_id;
-    sentry__ensure_event_id(event, &event_id);
+        SENTRY_DEBUG("capturing event");
 
-    SENTRY_WITH_SCOPE (scope) {
-        SENTRY_TRACE("merging scope into event");
-        sentry_scope_mode_t mode = SENTRY_SCOPE_ALL;
-        if (!opts->symbolize_stacktraces) {
-            mode &= ~SENTRY_SCOPE_STACKTRACES;
+        SENTRY_WITH_SCOPE (scope) {
+            SENTRY_TRACE("merging scope into event");
+            sentry_scope_mode_t mode = SENTRY_SCOPE_ALL;
+            if (!options->symbolize_stacktraces) {
```

Comment:
I'm afraid that we're locking too long in some of these functions. Particularly symbolicating the stack trace, or some of the backend functions can take considerable time. Sometimes, they even lock internally, again.

In a concurrent application, this could lead to significant delays. Would it be worth investigating alternative approaches here, for instance immutable options?